File: src/main/java/dev/esophose/discordbot/command/commands/info/FindRolesCommand.java
package dev.esophose.discordbot.command.commands.info;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.command.DiscordCommand;
import dev.esophose.discordbot.command.DiscordCommandMessage;
import dev.esophose.discordbot.manager.CommandManager;
import discord4j.core.object.entity.Guild;
import discord4j.core.object.entity.Role;
import discord4j.core.object.util.Permission;
import discord4j.core.object.util.PermissionSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import org.apache.commons.lang3.StringUtils;
import reactor.core.publisher.Mono;

public class FindRolesCommand extends DiscordCommand {

    public FindRolesCommand(Sparky bot) {
        super(bot);
    }

    public void execute(DiscordCommandMessage message, String input) {
        CommandManager commandManager = this.bot.getManager(CommandManager.class);

        message.getGuild()
                .flatMapMany(Guild::getRoles)
                .filter(x -> input.equals("*") || StringUtils.containsIgnoreCase(x.getName(), input))
                .sort(Comparator.comparingInt(Role::getRawPosition).reversed())
                .collectList()
                .switchIfEmpty(Mono.just(new ArrayList<>()))
                .flatMap(roles -> {
                    if (roles.isEmpty()) {
                        return commandManager.sendResponse(message.getChannel(), "No roles found", "No roles were found matching your input.");
                    } else {
                        StringBuilder stringBuilder = new StringBuilder();
                        roles.forEach(x -> stringBuilder.append(x.getMention()).append(" | ").append(x.getId().asString()).append('\n'));
                        return commandManager.sendResponse(message.getChannel(), roles.size() + " " + (roles.size() > 1 ? "Roles" : "Role") + " found matching \"" + input + "\"", stringBuilder.toString());
                    }
                })
                .subscribe();
    }

    @Override
    public String getName() {
        return "findroles";
    }

    @Override
    public List<String> getAliases() {
        return Collections.singletonList("fr");
    }

    @Override
    public String getDescription() {
        return "Finds roles matching the given input";
    }

    @Override
    public PermissionSet getRequiredBotPermissions() {
        return PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_ROLES);
    }

    @Override
    public Permission getDefaultRequiredMemberPermission() {
        return Permission.MANAGE_ROLES;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/commands/info/FindRolesCommand.kt
package dev.esophose.discordbot.command.commands.info

import dev.esophose.discordbot.Sparky
import dev.esophose.discordbot.command.DiscordCommand
import dev.esophose.discordbot.command.DiscordCommandMessage
import dev.esophose.discordbot.manager.CommandManager
import discord4j.core.`object`.entity.Role
import discord4j.rest.util.Permission
import discord4j.rest.util.PermissionSet
import org.apache.commons.lang3.StringUtils
import reactor.core.publisher.Mono
import java.util.*


class FindRolesCommand : DiscordCommand() {

    override val name: String
        get() = "findroles"

    override val aliases: List<String>
        get() = listOf("fr")

    override val description: String
        get() = "Finds roles matching the given input"

    override val requiredBotPermissions: PermissionSet
        get() = PermissionSet.of(Permission.SEND_MESSAGES, Permission.MANAGE_ROLES)

    override val defaultRequiredMemberPermission: Permission
        get() = Permission.MANAGE_ROLES

    fun execute(message: DiscordCommandMessage, input: String) {
        val commandManager = Sparky.getManager(CommandManager::class)

        message.guild
                .flatMapMany { it.roles }
                .filter { x -> input == "*" || StringUtils.containsIgnoreCase(x.name, input) }
                .sort(Comparator.comparingInt<Role> { it.rawPosition }.reversed())
                .collectList()
                .switchIfEmpty(Mono.just(ArrayList()))
                .flatMap { roles ->
                    if (roles.isEmpty()) {
                        commandManager.sendResponse(message.channel, "No roles found", "No roles were found matching your input.")
                    } else {
                        val stringBuilder = StringBuilder()
                        roles.forEach { x -> stringBuilder.append(x.mention).append(" | ").append(x.id.asString()).append('\n') }
                        commandManager.sendResponse(message.channel, roles.size.toString() + " " + (if (roles.size > 1) "Roles" else "Role") + " found matching \"" + input + "\"", stringBuilder.toString())
                    }
                }
                .subscribe()
    }

}
